### PR TITLE
Differential Testing and FFI

### PR DESCRIPTION
**Pull request by dmfxyz**
Added a write up on differential and standardized testing. Has examples of using `FFI` against non-evm executables, addressing #182.

Needs some editorial work, but content is about done so I am opening for review. It would be cool to include Frankie's GDA tests as an example as well, but avoiding doing that without permission/collaboration.

### CODE REVIEW

File: src/forge/differential-ffi-testing.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+function testMerkleGenerateProofStandard() public view {
+    bytes32[] memory _data = _getData(); 
+    for (uint i = 0; i < leaves.length; ++i) {
+        m.getProof(_data, leaves[i]);
+    }
+}
+```
+`src/test/standard_data/StandardInput.txt` is a text file that contains an encoded bytes32[100] array. It's generated outside of the test and can be used in any language's web3 sdk. It looks something like:
+
+```
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:118]: Tests are failing here, just add ```ignore

File: src/forge/differential-ffi-testing.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
+## Differential Testing
+
+Forge can enable differential fuzz testing between different implementations, and even against non-EVM executables using the `ffi` [cheatcode](../cheatcodes/ffi.md).
+
+[Differential testing](https://en.wikipedia.org/wiki/Differential_testing) cross references multiple implementations of the same function by comparing each one's output. Imagine we have a function specification `F(X)`, and two implementations of that specification: `f1(X)` and `f2(X)`. We expect `EQ(f1(x), f2(x))` for all x that exist in an appropriate input space and some equality function EQ. If `!EQ(f1(x), f2(x))`, we know that at least one function is incorrectly implementing `F(X)`. This process of testing for equality and identifying discrepancies is the core of differential testing.
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:5]: This is mostly a matter of habit for me, I'm not sure if the `EQ` notation is something formal - if it is, then feel free to skip

```suggestion
[Differential testing](https://en.wikipedia.org/wiki/Differential_testing) cross references multiple implementations of the same function by comparing each one's output. Imagine we have a function specification `F(X)`, and two implementations of that specification: `f1(X)` and `f2(X)`. We expect that `f1(x) == f2(x)` for all x that exist in an appropriate input space If `f1(x) != f2(x)`, then we know that at least one function is incorrectly implementing `F(X)`. This process of testing for equality and identifying discrepancies is the core of differential testing.
```

File: src/forge/differential-ffi-testing.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+Some real life uses include:
+* Comparing upgraded implementations to their predecessors
+* Testing code against known reference implementations
+* Confirming compatability with third party tools and dependencies
+
+Below are some examples on how you can use Foundry for differential testing.
+
+### Primer: The `ffi` cheatcode
+
+`ffi` allows you to execute an arbitrary shell command and capture the output. Here's a mock example:
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:16]: Just a little link to the FFI cheatcode

```suggestion
[`ffi`](../cheatcodes/ffi.md) allows you to execute an arbitrary shell command and capture the output. Here's a mock example:
```

File: src/forge/differential-ffi-testing.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        address loadedAddress = abi.decode(result, (address));
+        // Do something with the address
+        // ...
+    }
+}
+```
+An address has previously been written to `address.txt`, and we read it in using the cheatcode.
+
+### Example: Differential Testing Merkle Tree Implementations
+[Merkle Trees](https://en.wikipedia.org/wiki/Merkle_tree) are a cryptographic commitmentment scheme frequently used in blockchain applications. Their popularity means that there are a number of different implementations of Merkle Tree generators, provers, and verifiers. Often, Merkle roots and proofs are generated uisng a language like Javascript or Python, and proofs are verified on-chain in solidity.
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:37]: One typo + minor nits

```suggestion
[Merkle Trees](https://en.wikipedia.org/wiki/Merkle_tree) are a cryptographic commitmentment scheme frequently used in blockchain applications. Their popularity means that there are a number of different implementations of Merkle Tree generators, provers, and verifiers. Often, Merkle roots and proofs are generated using a language like JavaScript or Python, and proofs are verified on-chain in Solidity.
```

File: src/forge/differential-ffi-testing.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+[Murky](https://github.com/dmfxyz/murky) is a complete implementation of Merkle roots, proofs, and verification in solidity. Its test suite includes differential tests against OpenZeppelin's Merkle proof verification implementation, as well as root generation tests against a reference Javascript implementation. These tests are powered by Foundry's fuzzing and `ffi` capabilities.
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:39]:
```suggestion
[Murky](https://github.com/dmfxyz/murky) is a complete implementation of Merkle roots, proofs, and verification in Solidity. Its test suite includes differential tests against OpenZeppelin's Merkle proof verification implementation, as well as root generation tests against a reference JavaScript implementation. These tests are powered by Foundry's fuzzing and `ffi` capabilities.
```

File: src/forge/differential-ffi-testing.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+#### Differential Testing against reference TypeScript implementation
+Using the `ffi` cheatcode, Murky tests its own Merkle root implementation against a TypeScript implementation using data provided by forge's fuzzer:
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:42]:
```suggestion
Using the `ffi` cheatcode, Murky tests its own Merkle root implementation against a TypeScript implementation using data provided by Forge's fuzzer:
```

File: src/forge/differential-ffi-testing.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    bytes memory jsResult = vm.ffi(runJsInputs);
+    bytes32 jsGeneratedRoot = abi.decode(jsResult, (bytes32));
+    
+    // Calculate root using Murky
+    bytes32 murkyGeneratedRoot = m.getRoot(leaves);
+    assertEq(murkyGeneratedRoot, jsGeneratedRoot);
+}
+```
+
+forge runs `npm --prefix differential_testing/scripts/ --silent run generate-root-cli {numLeaves} {hexEncodedLeaves}`. This calculates the Merkle root for the input data using the reference JavaScript implementation. The script prints the root to stdout, and that printout is captured as `bytes` in the return value of `vm.ffi()`.
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:70]:
```suggestion
Forge runs `npm --prefix differential_testing/scripts/ --silent run generate-root-cli {numLeaves} {hexEncodedLeaves}`. This calculates the Merkle root for the input data using the reference JavaScript implementation. The script prints the root to stdout, and that printout is captured as `bytes` in the return value of `vm.ffi()`.
```

File: src/forge/differential-ffi-testing.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+The test then calculates the root using the solidity implementation.
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:72]:
```suggestion
The test then calculates the root using the Solidity implementation.
```

File: src/forge/differential-ffi-testing.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+Finally, the test asserts that the both roots are exactly equal. If not, the test fails.
+
+#### Differential testing against OpenZeppelin's Merkle Proof Library
+You may want to use differential testing against another solidity implementation. In that case, `ffi` is not needed. In this example, the reference implementation is imported directly into the test.
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:76]:
```suggestion
You may want to use differential testing against another Solidity implementation. In that case, `ffi` is not needed. In this example, the reference implementation is imported directly into the test.
```

File: src/forge/differential-ffi-testing.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    bytes32[] memory proof = m.getProof(_data, node);
+    bytes32 valueToProve = _data[node];
+    bool murkyVerified = m.verifyProof(root, proof, valueToProve);
+    bool ozVerified = MerkleProof.verify(proof, root, valueToProve);
+    assertTrue(murkyVerified == ozVerified);
+}
+```
+
+#### Standardized Testing against reference data
+FFI is also useful for injecting reproducible, standardized data into the testing environment. In the Murky library, this is used as a benchmark for gas snapshotting (see: [forge snapshot](./gas-snapshots.md)).
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:94]:
```suggestion
FFI is also useful for injecting reproducible, standardized data into the testing environment. In the Murky library, this is used as a benchmark for gas snapshotting (see [forge snapshot](./gas-snapshots.md)).
```

File: src/forge/differential-ffi-testing.md [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+}
+    
+function testMerkleGenerateProofStandard() public view {
+    bytes32[] memory _data = _getData(); 
+    for (uint i = 0; i < leaves.length; ++i) {
+        m.getProof(_data, leaves[i]);
+    }
+}
+```
+`src/test/standard_data/StandardInput.txt` is a text file that contains an encoded bytes32[100] array. It's generated outside of the test and can be used in any language's web3 sdk. It looks something like:
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:116]:
```suggestion
`src/test/standard_data/StandardInput.txt` is a text file that contains an encoded `bytes32[100]` array. It is generated outside of the test and can be used in any language's Web3 SDK. It looks something like:
```

**Review by onbjerg — CHANGES_REQUESTED** [review]
Amazing! Very simple to understand, at least in my opinion - I had not come across differential testing before, but the concept seems simple and I understood what was conveyed in the text

I've left some very small nits, most of them are just some casing/links and 1 typo

**Comment by FrankieIsLost** [comment]
I apologize for the slow response here, this look great! Really thorough write-up and examples, and lines up with how I've been using differential testing. 

One minor suggestion might be to clarify the distinction between differential testing and differential fuzzing. For example, you can write differential tests with FFI without relying on the fuzzer, which is helpful when you want to manually check certain edge-cases. 

Regarding GDA testing, feel free to link. 

Thanks for taking point on this!

File: src/forge/differential-ffi-testing.md (hunk 2 above)
Comment (dmfxyz) [reply src/forge/differential-ffi-testing.md]: Ack. I used `EQ` because it's not always a strict equivalence function the way many programmers may think when they see `==`, but I replaced it with your suggestion and just added a note as I do think it's more readable.

**Comment by dmfxyz** [comment]
Thanks for the feedback! @onbjerg I addressed your first round of comments. Agree that I should add a non-fuzzed differential use-case, and I already have an example of that in the Murky repo. I'll write that up tonight and also link to the GDA repo, and then I think we should be ready for another review and merge.

**Comment by dmfxyz** [comment]
OK! I think we're ready for a final review. cc @gakonst as well if you have time to give it a quick read.

File: src/forge/differential-ffi-testing.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+Differential fuzzing is an extension of differential testing. Differential fuzzing programatically generates many values of `x` to find discrepencies and edge cases that manually chosen inputs might not reveal. 
+
+> Note: the `==` operator in this case can be a custom definition of equality. For example, if testing floating point implementations, you might use approximate equality with a certain tolerance.
+
+Some real life uses of uses of this type of testing include:
+* Comparing upgraded implementations to their predecessors
+* Testing code against known reference implementations
+* Confirming compatability with third party tools and dependencies
```
Comment (gakonst) [inline src/forge/differential-ffi-testing.md:15]:
```suggestion
* Confirming compatibility with third party tools and dependencies
```

**Review by gakonst — APPROVED** [review]

File: src/forge/differential-ffi-testing.md [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+> Of course, one could just hardcode the array into the test! But that makes it much harder to do consistent testing across contracts, implementations, etc.
+
+### Example: Differential Testing Gradual Dutch Auctions
+The reference implementation for Paradigm's [Gradual Dutch Auction](https://www.paradigm.xyz/2022/04/gda) mechanism contains a number of differential, fuzzed tests. It is an excellent repository to further explore differential testing using `ffi`.
+
+* Differential tests for [Discrete GDAs](https://github.com/FrankieIsLost/gradual-dutch-auction/blob/master/src/test/DiscreteGDA.t.sol#L78)
+  
+* Differential tests for [Continuous GDAs](https://github.com/FrankieIsLost/gradual-dutch-auction/blob/master/src/test/ContinuousGDA.t.sol#L89)
+
+* Reference [Python implementation](https://github.com/FrankieIsLost/gradual-dutch-auction/blob/master/analysis/compute_price.py)
```
Comment (onbjerg) [inline src/forge/differential-ffi-testing.md:143]: Spacing is off in the rendered doc otherwise

```suggestion
* Differential tests for [Discrete GDAs](https://github.com/FrankieIsLost/gradual-dutch-auction/blob/master/src/test/DiscreteGDA.t.sol#L78)
* Differential tests for [Continuous GDAs](https://github.com/FrankieIsLost/gradual-dutch-auction/blob/master/src/test/ContinuousGDA.t.sol#L89)
* Reference [Python implementation](https://github.com/FrankieIsLost/gradual-dutch-auction/blob/master/analysis/compute_price.py)
```

**Review by onbjerg — APPROVED** [review]
last nit and then we merge!!

**Comment by mds1** [comment]
Wow this is really great, nice and detailed!! Awesome job @dmfxyz

**Comment by cruzdanilo** [comment]
@dmfxyz  where does `(bytes memory).toHexString()` come from?

**Comment by dmfxyz** [comment]
> @dmfxyz where does `(bytes memory).toHexString()` come from?

@cruzdanilo this is from library implementation I wrote in for the Murky library. I called it [`Strings2]`(https://github.com/dmfxyz/murky/blob/main/differential_testing/test/utils/Strings2.sol) for now. Might move it out of that repo to somewhere more shared, but feel free to copy paste as you need for now.
